crucible-llvm/src/Lang/Crucible/LLVM/SymIO.hs

```diff
@@ -524,3 +540,21 @@ are here, but they are arranged to not collide with any valid names for actual
 files in the symbolic filesystem.
 
 -}
+
+{- Note [File Descriptor Sequence]
```

Comment:
The documentation for `dup` and such indicates that the order in which fds are allocated is pretty deterministic.  Real code can (although usually doesn't) depend on that behavior.  I think this design is fine for most purposes, but we should probably start an "errata" list or something.

Other designs we could consider:
1) Accurately following UNIX behavior (allocating the smallest unused fd)
2) Protect the abstraction further by allocating fresh non-dereferencable pointers instead of actually using small integers.

2) is even less accurate than what we are doing now, but is more aggressive about maintaining the abstraction.

Reply:
I was thinking about the implications of just making the file descriptor source be symbolic so that it could take part in branching and merging. I'm not sure what the implications would be...

As for 2, I think it is noted somewhere but we would have to build an escape hatch to support stdin/stdout/stderr. That isn't necessarily a problem, but just another consideration.

Reply:
Well, it couldn't just be the FD source.  You'd actually need to track which FDs have been closed and such, because they get recycled.